sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -975,6 +939,73 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
       } else {
         filter
       }
+
+    case filter @ Filter(condition, u: Union) =>
```

Comment:
yea I agree, but this depends on the current implementation of `Union`, and it's weird to use the RDD partition stuff while working on SQL optimization rules...